projects/step-frontend/src/lib/modules/plan-editor/components/plan-editor-base/plan-editor-base.component.ts

```diff
@@ -154,19 +167,42 @@ export class PlanEditorBaseComponent
   }
 
   displayHistory(permission: string, plan: Plan): void {
-    if (!plan || !plan.id) {
+    if (!(plan?.id || this.id)) {
       return;
     }
 
-    const planVersion = plan.customFields ? plan.customFields['versionId'] : undefined;
-
     const versionHistory = this._planEditorApi.getPlanHistory(this.id || plan.id!);
 
-    this._restoreDialogsService
-      .showRestoreDialog(planVersion, versionHistory, permission)
+    const currentVersion$ = defer(() => {
+      if (this.id && this.id !== plan.id) {
+        // composite keywords need to retrieve the current version
+        return this._keywordCallsApi.getFunctionById(this.id).pipe(
+          mergeMap((keyword) => {
+            if (keyword && keyword.customFields && keyword.customFields['versionId']) {
+              return of(keyword.customFields['versionId']);
+            } else {
+              return EMPTY;
+            }
+          })
+        );
+      } else {
+        // we are showing a real plan
+        return of(plan.customFields ? plan.customFields['versionId'] : undefined);
+      }
+    });
+
+    currentVersion$
       .pipe(
-        filter((restoreVersion) => !!restoreVersion),
-        switchMap((restoreVersion) => this._planEditorApi.restorePlanVersion(this.id || plan.id!, restoreVersion))
+        mergeMap((currentVersion) =>
+          this._restoreDialogsService.showRestoreDialog(currentVersion, versionHistory, permission).pipe(
+            filter((restoreVersion) => restoreVersion !== undefined),
+            mergeMap((restoreVersion) => this._planEditorApi.restorePlanVersion(this.id || plan.id!, restoreVersion)),
+            catchError((error) => {
+              console.error(error);
+              return EMPTY;
+            })
+          )
+        )
```

Comment:
Looks like that inner `pipe` is redundant.
Also I prefer to use the `switchMap` instead of `mergeMap`, because `switchMap` guarantees that previous observable has been completed. Here it's not significant because we have several sequenced requests, but in some cases `mergeMap` might be the source of memory leaks (ex. If the request is performed on each user input. With mergeMap each new request won't complete the previous one. Plus If the result observable will be shared for several subscriptions, there might be issues)

```suggestion
        switchMap((currentVersion) => this._restoreDialogsService.showRestoreDialog(currentVersion, versionHistory, permission)),
        filter((restoreVersion) => restoreVersion !== undefined),
        switchMap((restoreVersion) => this._planEditorApi.restorePlanVersion(this.id || plan.id!, restoreVersion)),
        catchError((error) => {
            console.error(error);
            return EMPTY;
        })
```